File: src/leetcode/m787/CheapestFlightsWithinKStops.java
/*
 * Copyright (c) 2024-2025 THGEEK.COM. All rights reserved.
 */

package leetcode.m787;

import java.util.Arrays;

/**
 * LeetCode 787
 *
 * @author Tao Hong
 * @version 1.0
 * @since 2024/07/26 15:11
 */
public class CheapestFlightsWithinKStops {
    public int findCheapestPrice(int n, int[][] flights, int src, int dst, int k) {
        int[] cost = new int[n];
        Arrays.fill(cost, Integer.MAX_VALUE);
        cost[src] = 0;

        for (int i = 0; i <= k; i++) {
            int[] temp = Arrays.copyOf(cost, n);
            for (int[] flight : flights) {
                int u = flight[0];
                int v = flight[1];
                int w = flight[2];
                if (cost[u] != Integer.MAX_VALUE && temp[v] > cost[u] + w) {
                    temp[v] = cost[u] + w;
                }
            }
            cost = temp;
        }

        return cost[dst] == Integer.MAX_VALUE ? -1 : cost[dst];
    }
}

File: src/leetcode/m787/CheapestFlightsWithinKStops.go
package m787

import "math"

// LeetCode 787
func findCheapestPrice(n int, flights [][]int, src int, dst int, k int) int {
	costs := make([]int, n)
	for i := range costs {
		costs[i] = math.MaxInt
	}
	costs[src] = 0

	for i := 0; i <= k; i++ {
		temp := make([]int, n)
		copy(temp, costs)
		for _, flight := range flights {
			f, t, p := flight[0], flight[1], flight[2]
			if costs[f] != math.MaxInt && temp[t] > costs[f]+p {
				temp[t] = costs[f] + p
			}
		}
		costs = temp
	}

	if costs[dst] == math.MaxInt {
		return -1
	}
	return costs[dst]
}
